### ISSUE TITLE
GTK4 Port

### ISSUE THREAD

**Issue by LaserEyess** [transmission-remote-gtk/transmission-remote-gtk#117]
Is there any intention to port this to gtk4 or handle potential bitrotting? I'm looking through the [porting guide](https://developer.gnome.org/gtk4/stable/gtk-migrating-3-to-4.html) and at the first step of porting, compiling with `make CFLAGS+="-DGDK_DISABLE_DEPRECATED -DGTK_DISABLE_DEPRECATED"` is giving a massive amount of errors.

https://0x0.st/-qfU.log

I'm sure this is a lot of work, but this program is pretty important to my workflow so I'd be interested in helping.

**Comment by TingPing** [comment]
Nobody is working on it. The project could probably use a new maintainer at this point.

**Comment by ajf8** [comment]
Indeed, a new maintainer or maintainers would be much appreciated. I would like to see the project continue and can possibly help and support that occasionally. I'm also happy to test/review/merge PR's (especially GTK4), but I don't really use bittorrent or the Linux desktop much these days.

Would be a shame for the project to die after over 10 years (I created it over Christmas 2010). I don't think it's aged too badly all things considered! But yeah it does need a bit of attention. We should get a release out, as there's been a few useful PRs over time.

**Comment by LaserEyess** [comment]
As I said I would be interested in helping, but my knowledge of C is too limited to make significant contributions to a project of this size. Furthermore I'm not too familiar with glib2 or gtk programming. That being said, I can help with some handholding, as well as with testing.

**Comment by TingPing** [comment]
I think its a decent chunk of work for somebody new to the platform. Not to say its impossible if you are motivated.

I *think* the first step of removing deprecated usage is probably doable. Its mostly documented and plenty of example code out there from other projects. I would enable deprecations, fix compiling a single broken `.c` file at a time, then disable deprecations making sure everything works together, and repeat until the entire project builds.

The step of moving to GTK4 would require larger refactors. For example all tree view usage must replace cell renderers with real widgets.

Just speaking for myself I don't have much time for handholding.

**Comment by LaserEyess** [comment]
Starting to look into this, and immediately I've run into some massive design changes that seem to be unavoidable. The first step has been trying to replace the deprecated `GTK_STOCK_*` defines with their actual label/icon names. See [here](https://docs.google.com/spreadsheets/u/0/d/1HavJQRPpMuq-N0GoN1wJR-9KEGXpKy3-NEPpZZkUGJY/pub?output=html), (apologies for the google docs link. I actually cannot find this on the gtk docs site). This change, effectively, destroys all menu icons. I do not see any way to get them back, and based on [this](https://igurublog.wordpress.com/2014/03/22/gtk-3-10-drops-menu-icons-and-mnemonics/) this is completely intended.

My question is: when following guides to remove deprecation, what level of feature "breakage" is acceptable? In particular, replacing the `GTK_STOCK_*` stuff is a lot of work, both monotonous sed'ing, as well as changing up a few functions, so I'd like to do it correctly the first time.

**Comment by TingPing** [comment]
> This change, effectively, destroys all menu icons. I do not see any way to get them back, and based on this this is completely intended.

You *can* manually pack an icon into a menu, it just isn't the same exposed API. My personal opinion is menus shouldn't have icons any longer, its a design that has fallen out of favor.

> My question is: when following guides to remove deprecation, what level of feature "breakage" is acceptable?

I think this is entirely reasonable and STOCK_ICONS have been deprecated for like a decade.

**Comment by LaserEyess** [comment]
Thanks for the advice. Seems pretty clear that in this case factoring out stock icons is not only good practice but also a hard requirement. So I'll probably get a PR for this within the week, assuming I don't get bored of `sed`.
